doc/merkle.md

```diff
@@ -0,0 +1,139 @@
+# Merkle Structures
+
+MimbleWimble is designed for users to verify the state of the system given
+only pruned data. To achieve this goal, all transaction data is committed
+to the blockchain by means of Merkle trees which should support efficient
+updates and serialization even when pruned.
+
+Also, almost all transaction data (inputs, outputs, excesses and excess
+proofs) have the ability to be summed in some way, so it makes sense to
+treat Merkle sum trees as the default option, and address the sums here.
+
+A design goal of Grin is that all structures be as easy to implement and
+as simple as possible. MimbleWimble introduces a lot of new cryptography
+so it should made as easy to understand as possible. Its validation rules
+are simple to specify (no scripts) and Grin is written in a language with
+very explicit semantics, so simplicity is also good to achieve well-understood
+consensus rules.
+
+## Merkle Trees
+
+There are three Merkle trees committed to by each block:
+
+### Total Output Set
+
+Each object is one of two things: a (commitment, rangeproof) pair indicating
+an unspent output or a NULL marker indicating a spent one. It is a sum-tree
+over all unspent outputs (spent ones contribute nothing to the sum).
```

Comment:
An output without a rangeproof is 32 bytes. Ten millions of them is only 320MB. Ten million outputs with the rangeproof (fully blinding) is over 50 GB.

I see the UTXO Merkle tree as a specialized index. Updating it and recomputing the root is going to be the most expensive validation operation. Someone will likely try to DDoS it at some point. Without the rangeproof it easily fits in memory, gives us fast block processing and fast reply to a good range of queries, and makes a node less susceptible to random reads types of attacks. Even with a billion output, because of the use of MMR, we can still fit the most recent parts in memory.

Is see the rangeproof as reference data. It will fit on disk with transactions and blocks, likely part of a nice full sequential write and rarely read, may be eventually pruned based on node policies, etc.

Makes sense?